src/ZVal.h

```diff
@@ -0,0 +1,396 @@
+// given we have a union and thus on destruction C++ doesn't know which
+// member flavor to destruct.
+//
+// Note that a ZAMValUnion by itself is ambiguous: it doesn't track its
+// type.  This makes them consume less memory and cheaper to copy.  It
+// does however require a separate way to determine the type.  Generally
+// this is doable using surrounding context, or can be statically determined
+// in the case of optimization/compilation.
+//
+// An alternative would be to use std::variant, but (1) it tracks the
+// variant type, and (2) it won't allow access to the managed_val member,
+// which not only simplifies memory management but also is required for
+// sharing of ZAM frame slots.
+union ZAMValUnion {
+	// Constructor for hand-populating the values.
+	ZAMValUnion() { managed_val = nullptr; }
+
+	// Construct from a given Bro value with a given type.
+	ZAMValUnion(zeek::IntrusivePtr<zeek::Val> v,
+			const zeek::IntrusivePtr<zeek::Type>& t);
+
+	// True if when interpreting the value as having the given type,
+	// it's a nil pointer.
+	bool IsNil(const zeek::IntrusivePtr<zeek::Type>& t) const;
+
+	// Convert to a Bro value.
+	zeek::IntrusivePtr<zeek::Val>
+		ToVal(const zeek::IntrusivePtr<zeek::Type>& t) const;
+
+	// Used for bool, int, enum.
+	bro_int_t int_val;
+
+	// Used for count, counter, port.
+	bro_uint_t uint_val;
+
+	// Used for double, time, interval.
+	double double_val;
+
+	// The types are all variants of Val, zeek::Type, or more fundamentally
+	// zeek::Obj.  They are raw pointers rather than IntrusivePtr's because
+	// unions can't contain the latter.  For memory management, we use
+	// Ref/Unref.
+	zeek::StringVal* string_val;
+	zeek::AddrVal* addr_val;
+	zeek::SubNetVal* subnet_val;
+	zeek::File* file_val;
+	zeek::Func* func_val;
+	zeek::ListVal* list_val;
+	zeek::OpaqueVal* opaque_val;
+	zeek::PatternVal* re_val;
+	zeek::TableVal* table_val;
+	zeek::RecordVal* record_val;
+	zeek::VectorVal* vector_val;
+	zeek::Type* type_val;
+
+	// Used for direct "any" values.
+	zeek::Val* any_val;
+
+	// Used for the compiler to hold opaque items.  Memory management
+	// is explicit in the operations accessing it.
+	val_vec* vvec;
+
+	// Used by the compiler for managing "for" loops.  Implicit
+	// memory management.
+	IterInfo* iter_info;
+
+	// Used for generic access to managed (reference-counted) objects.
+	zeek::Obj* managed_val;
+};
+
+// True if a given type is one for which we manage the associated
+// memory internally.
+bool IsManagedType(const zeek::IntrusivePtr<zeek::Type>& t);
+
+// Deletes a managed value.
+inline void DeleteManagedType(ZAMValUnion& v)
+	{
+	Unref(v.managed_val);
+	}
+
+// The following can be set to point to a boolean that will be set
+// to true if a run-time error associated with ZAMValUnion's occurs.
+//
+// We use this somewhat clunky coupling to enable isolating ZVal from
+// ZAM compiler specifics.
+extern bool* zval_error_addr;
```

Comment:
I assume this will mostly be needed for the compiler branch? In the current PR, I only found two uses for it, and it never seems to be read.

Reply:
Right, that's just needed for ZAM code execution, to catch run-time errors.